arch/riscv/core/semihost.c

```diff
@@ -5,9 +5,20 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include <zephyr/toolchain.h>
 #include <zephyr/arch/common/semihost.h>
 
-long semihost_exec(enum semihost_instr instr, void *args)
+/*
+ * QEMU requires that the semihosting trap instruction sequence, consisting of
```

Comment:
Not sure why blame QEMU if this info is also at https://github.com/riscv/riscv-semihosting-spec/blob/main/riscv-semihosting-spec.adoc.

Reply:
> On systems with paging support, this sequence must not cross a page boundary as the semihosting system must be able to check for the semihosting sequence without needing data from potentially missing pages.

because of the "On systems with paging support" part

Reply:
Yeah, sorry for the mismatch between the spec and the qemu implementation. I started with the qemu implementation; during early review for that the issue with page boundaries came up. Merging was blocked without an "official" spec, so I wrote that based on the design done by the other authors. I realized that a page-boundary restriction didn't make much sense while writing that section, so only required it on MMU systems. By that time, the qemu implementation was heading through review and I didn't want to mess with the code. If someone wants to provide a patch to qemu to follow the spec more closely, they might be interested in accepting it...